File: scala/src/main/scala/ar/uba/dc/cap/Term.scala
package ar.uba.dc.cap

import ar.uba.dc.cap._

object Term {
  type Branch = (Pattern, cap.Ctx, Term)
}

abstract class Term {
  def $(other: Term) : Term = {
    App(this, other)
  }
}
case class Var(label: String) extends Term
case class Const(label: String) extends Term
case class App(left: Term, right: Term) extends Term
case class Case(branches: Array[Term.Branch]) extends Term

abstract class Pattern {
  def freeMatchables() : List[cap.Label] = this match {
    case PMatch(label) => List(label)
    case PConst(label) => List()
    case PComp(left, right) => left.freeMatchables() ++ right.freeMatchables()
  }
  
  def $(other: Pattern) : Pattern = {
    PComp(this, other)
  }
}
case class PMatch(label: String) extends Pattern
case class PConst(label: String) extends Pattern
case class PComp(left: Pattern, right: Pattern) extends Pattern

File: haskell/src/Term.hs
module Term where

import Util
import Type
import qualified Data.Map.Strict as Map

data Term = Var Label
          | Const Label
          | App Term Term
          | Case [Branch]
          deriving Show

data Pattern = PMatch Label
             | PConst Label
             | PComp Pattern Pattern
             deriving (Eq, Show)

type Branch     = (Pattern, Ctx, Term)
type Assignment = (Pattern,Type)
type Ctx        = Map.Map Label Type

foldPattern :: (Label -> b)
               -> (Label -> b)
               -> (b -> b -> b)
               -> Pattern
               -> b
foldPattern fMatchable fConst fComp p = let rec = foldPattern fMatchable fConst fComp
                                        in case p of
                                            PMatch l -> fMatchable l
                                            PConst l    -> fConst l
                                            PComp p1 p2 -> fComp (rec p1) (rec p2)